## Comparative genomics and association analysis to identify genetic attributes associated with bacterial lifestyle

This repository provides supporting information and the codes for the following manuscript available as preprint in bioRXiv (doi: https://doi.org/10.1101/2023.05.31.543148) and is under peer review. 

### Genetic and functional diversity help explain pathogenic, weakly pathogenic, and commensal lifestyles in the genus *Xanthomonas*
Michelle M. Pena<sup>*$1</sup>, Rishi Bhandari<sup>*1</sup>, Robert M. Bowers<sup>5</sup>, Kylie Weis<sup>1</sup>, Eric Newberry<sup>1</sup>, Naama Wagner<sup>2</sup>, Tal Pupko<sup>2</sup>, Jeffrey B. Jones<sup>4</sup>, Tanja Woyke<sup>5,6</sup>, Boris A. Vinatzer<sup>3</sup>, Marie- Agnès Jacques<sup>7</sup>, Neha Potnis<sup>1#</sup>

<sup>1</sup> Department of Entomology and Plant Pathology, Auburn University, Auburn, AL, USA
<sup>2</sup> The Shmunis School of Biomedicine and Cancer Research,Israel
<sup>3</sup> School of Plant and Environmental Sciences, Virginia Tech, Blacksburg, VA, USA 
<sup>4</sup> Department of Plant Pathology, University of Florida, Gainesville, FL, USA
<sup>5</sup> DOE Joint Genome Institute, Lawrence Berkeley National Laboratory, Berkeley, CA, USA
<sup>6</sup> Life and Environmental Sciences, University of California Merced, Merced, CA, USA 
<sup>7</sup> Univ Angers, Institut Agro, INRAE, IRHS, SFR QUASAV, F-49000 Angers, France

<sup>#</sup>Corresponding author: Neha Potnis, nzp0024@auburn.edu

#### <sup>*</sup>Michelle M. Pena and Rishi Bhandari contributed equally to this manuscript.



The manuscript has been submitted for peer review and is available in *bioRxiv* 


## Abstract
The genus *Xanthomonas* has been primarily studied for pathogenic interactions with plants. However, besides host and tissue specific pathogenic strains, this genus also comprises nonpathogenic strains isolated from a broad range of hosts, sometimes in association with pathogenic strains, and other environments, including rainwater. Based on their incapacity or limited capacity to cause symptoms on the host of isolation, nonpathogenic xanthomonads can be further characterized as commensal and weakly pathogenic. This study aimed to understand the diversity and evolution of nonpathogenic xanthomonads compared to their pathogenic counterparts based on their co-occurrence and phylogenetic relationship and to identify genomic traits that form the basis of a life-history framework that groups xanthomonads by ecological strategies. We sequenced genomes of 83 strains spanning the genus phylogeny and identified eight novel species, indicating unexplored diversity. While some nonpathogenic species have experienced a recent loss of a type III secretion system, specifically, the *hrp2* cluster, we observed an apparent lack of association of the hrp2 cluster with lifestyles of diverse species. We gathered evidence for gene flow among co-occurring pathogenic and nonpathogenic strains, suggesting the potential of nonpathogenic strains to act as a reservoir of adaptive traits for pathogenic strains and vice versa. We further identified traits enriched in nonpathogens that suggest a strategy of stress tolerance, rather than avoidance, during their association with a broad range of host plants.


## Data Availability

Sequence data generated from this work have been deposited in the DOE JGI and GenBank database under the JGI taxon id and Bioproject accession number as given in the metadata table.


## Funding
The work (proposal: 10.46936/10.25585/60001156) conducted by the US Department of Energy Joint Genome Institute (https://ror.org/04xm1d337), a DOE Office of Science User Facility, is supported by the Office of Science of the US Department of Energy operated under Contract No. DE-AC02-05CH11231.
We thank CIRM-CFBP (Beaucouzé, INRAE, France http://www6.inra.fr/cirm_eng/CFBP -Plant- Associated-Bacteria) for strain preservation and supply. We also thank Abi S. A. Marques and Marisa A. S. V. Ferreira for participating in the strain sampling from bean.
